From e049e4b4809dea3674cd3f6cba38c609d5c695a6 Mon Sep 17 00:00:00 2001
From: Orgad Shaneh <orgad.shaneh@audiocodes.com>
Date: Mon, 4 May 2026 20:34:20 +0300
Subject: [PATCH 1/2] MT#55283 preserve supplemental codecs in publish mode

Add RED (RFC 2198) codec definition to codeclib as a supplemental codec.
Modify codec_store_accept_one() to preserve supplemental codecs whose
clock rate matches the accepted primary codec. Skip supplemental codecs
in codec_store_is_full_answer() since they are optional in SDP answers.

Closes #2100

Change-Id: Ie81f45091c5d3dc55e20896d5a6e7a8361fda2f4
(cherry picked from commit b87ae5803642828ed8e7a18f8316361fc0b74b98)
---
 daemon/codec.c | 12 +++++++++++-
 lib/codeclib.c | 12 ++++++++++++
 2 files changed, 23 insertions(+), 1 deletion(-)

diff --git a/daemon/codec.c b/daemon/codec.c
index f081d477..2d93d840 100644
--- a/daemon/codec.c
+++ b/daemon/codec.c
@@ -6051,7 +6051,8 @@ int codec_store_accept_one(struct codec_store *cs, str_q *accept, bool accept_an
 		return -1;
 	}
 
-	// delete all codecs except the accepted one
+	// delete all codecs except the accepted one, preserving supplemental codecs
+	// that match the accepted codec's clock rate
 	__auto_type link = cs->codec_prefs.head;
 	while (link) {
 		rtp_payload_type *pt = link->data;
@@ -6059,6 +6060,13 @@ int codec_store_accept_one(struct codec_store *cs, str_q *accept, bool accept_an
 			link = link->next;
 			continue;
 		}
+		ensure_codec_def(pt, cs->media);
+		if (pt->codec_def && pt->codec_def->supplemental
+				&& pt->clock_rate == accept_pt->clock_rate)
+		{
+			link = link->next;
+			continue;
+		}
 		link = __codec_store_delete_link(link, cs);
 	}
 
@@ -6372,6 +6380,8 @@ void codec_store_synthesise(struct codec_store *dst, struct codec_store *opposit
 bool codec_store_is_full_answer(const struct codec_store *src, const struct codec_store *dst) {
 	for (auto_iter(l, src->codec_prefs.head); l; l = l->next) {
 		const rtp_payload_type *src_pt = l->data;
+		if (src_pt->codec_def && src_pt->codec_def->supplemental)
+			continue;
 		const rtp_payload_type *dst_pt = t_hash_table_lookup(dst->codecs,
 				GINT_TO_POINTER(src_pt->payload_type));
 		if (!dst_pt || !rtp_payload_type_eq_compat(src_pt, dst_pt)) {
diff --git a/lib/codeclib.c b/lib/codeclib.c
index 0ade66d3..7bc3dd10 100644
--- a/lib/codeclib.c
+++ b/lib/codeclib.c
@@ -773,6 +773,18 @@ static struct codec_def_s __codec_defs[] = {
 		.format_cmp = format_cmp_ignore,
 		.codec_type = &codec_type_cn,
 	},
+	{
+		.rtpname = "red",
+		.avcodec_id = -1,
+		.packetizer = packetizer_passthrough,
+		.media_type = MT_AUDIO,
+		.supplemental = 1,
+		.default_clockrate = 8000,
+		.default_channels = 1,
+		.format_cmp = format_cmp_ignore,
+		.support_encoding = 1,
+		.support_decoding = 1,
+	},
        {
                .rtpname = "G726-16",
                        .avcodec_id = AV_CODEC_ID_ADPCM_G726,
-- 
2.53.0

